self_supervised/vision/swav.py

```diff
@@ -0,0 +1,172 @@
+        sum_Q = torch.sum(Q)
+        Q /= sum_Q
+
+        r = (torch.ones(Q.shape[0]) / Q.shape[0]).to(device)
+        c = (torch.ones(Q.shape[1]) / Q.shape[1]).to(device)
+
+        curr_sum = torch.sum(Q, dim=1)
+
+        for it in range(nmb_iters):
+            u = curr_sum
+            Q *= (r / u).unsqueeze(1)
+            Q *= (c / torch.sum(Q, dim=0)).unsqueeze(0)
+            curr_sum = torch.sum(Q, dim=1)
+        return (Q / torch.sum(Q, dim=0, keepdim=True)).t().float()
+
+
+    @torch.no_grad()
+    def _dequeue_and_enqueue(self, embedding):
+        assert self.K % self.bs == 0  # for simplicity
+        self.queue[self.queue_ptr:self.queue_ptr+self.bs, :] = embedding
+        self.queue_ptr = (self.queue_ptr + self.bs) % self.K  # move pointer
+
+
+    @torch.no_grad()
+    def _compute_codes(self, output):
+        qs = []
+        for i in self.crop_assgn_ids:
+            # use queue
+            if (self.K is not None) and (self.learn.pct_train > self.queue_start_pct):
+                target_b = output[self.bs*i:self.bs*(i+1)]
+                queue_b = self.learn.model.prototypes(self.queue)
+                merged_b = torch.cat([target_b, queue_b])
+                q = torch.exp(merged_b/self.eps).t()
+                q = self.sinkhorn_knopp(q, self.n_sinkh_iter, q.device)
+                qs.append(q[:self.bs])
+
+            # don't use queue
+            else:
+                target_b = output[self.bs*i:self.bs*(i+1)]
+                q = torch.exp(target_b/self.eps).t()
+                q = self.sinkhorn_knopp(q, self.n_sinkh_iter, q.device)
+                qs.append(q)
+        return qs
+
+
+    def after_pred(self):
+        "Compute ps and qs"
+
+        embedding, output = self.pred
+
+        # Update - no need to store all assignment crops, e.g. just 0 from [0,1]
+        # Update queue only during training
+        if (self.K is not None) and (self.learn.training):  self._dequeue_and_enqueue(embedding[:self.bs])
+
+        # Compute codes
+        qs = self._compute_codes(output)
+
+        # Compute predictions
+        log_ps = []
+        for v in np.arange(len(self.augs)):
+            log_p = F.log_softmax(output[self.bs*v:self.bs*(v+1)] / self.temp, dim=1)
+            log_ps.append(log_p)
+
+        log_ps, qs = torch.stack(log_ps), torch.stack(qs)
+        self.learn.pred, self.learn.yb = log_ps, (qs,)
+
+
+    def lf(self, pred, *yb):
+        log_ps, qs, loss = pred, yb[0], 0
+        t = (qs.unsqueeze(1)*log_ps.unsqueeze(0)).sum(-1).mean(-1)
+        for i, ti in enumerate(t): loss -= (ti.sum() - ti[i])/(len(ti)-1)/len(t)
+        return loss
+
+    @torch.no_grad()
+    def show_one(self):
+        xb = self.learn.xb[0]
+        i = np.random.choice(self.bs)
+        images = [aug.decode(b.to('cpu').clone()).clamp(0.1)[i]
+                      for b, aug in zip(xb, self.augs)]
+        return show_batch(xb[0], None, images, max_n=len(images), ncols=len(images), nrows=1)
```

Comment:
I would like to unify the `show` interface.
So I would rather have a single `show` method with a custom number of return images, just like in the SimCLR case :)
Here is my implementation:
```python
def show(self, n=1):
    xbs = self.learn.xb[0]
    idxs = np.random.choice(range(self.bs), n, False)
    images = [aug.decode(xb.to('cpu').clone()).clamp(0, 1)[i] 
              for i in idxs
              for xb, aug in zip(xbs, self.augs)]
    return show_batch(images[0], None, images, nrows=n)
```

While we are already working on integrating all of my custom changes ;D

Reply:
Sure, why not :D